src/routes/safe/components/Apps/communicator.ts

```diff
@@ -0,0 +1,103 @@
+import { MutableRefObject, useEffect, useState } from 'react'
+import {
+  getSDKVersion,
+  SDKMessageEvent,
+  MethodToResponse,
+  Methods,
+  ErrorResponse,
+  MessageFormatter,
+} from '@gnosis.pm/safe-apps-sdk'
+import { SafeApp } from './types.d'
+
+type MessageHandler = (
+  msg: SDKMessageEvent,
+) => void | MethodToResponse[Methods] | ErrorResponse | Promise<MethodToResponse[Methods] | ErrorResponse | void>
+
+class AppCommunicator {
+  private iframe: HTMLIFrameElement
+  private handlers = new Map<Methods, MessageHandler>()
+  private app: SafeApp
+
+  constructor(iframeRef: MutableRefObject<HTMLIFrameElement>, app: SafeApp) {
+    this.iframe = iframeRef.current
+    this.app = app
+
+    window.addEventListener('message', this.handleIncomingMessage)
+  }
+
+  on = (method: Methods, handler: MessageHandler): void => {
+    this.handlers.set(method, handler)
+  }
+
+  private isValidMessage = (msg: SDKMessageEvent): boolean => {
+    const sameOrigin = msg.origin === window.origin
```

Comment:
This causes issues if both the safe app and the interface using it are hosted on the same host (e.g. if both would be hosted via ipfs). We should check `msg.source === window`

Reply:
perhaps, renaming it to `isValidOrigin`?  `isValidMessage` should also validate if the method is known.

Reply:
I named it this way because in the future it may have more checks, not only the origin. Good catch with the method validation.